forms/CountryDropdownField.php

```diff
@@ -28,48 +33,57 @@ class CountryDropdownField extends DropdownField {
 	 * @return string
 	 */
 	protected function locale() {
-		if (($member = Member::currentUser()) && $member->Locale) return $member->Locale;
+		if (($member = Member::currentUser()) && $member->Locale) {
+			return $member->Locale;
+		}
 		return i18n::get_locale();
 	}
 
-	public function __construct($name, $title = null, $source = null, $value = "", $form=null) {
-		if(!is_array($source)) {
-			// Get a list of countries from Zend
-			$source = Zend_Locale::getTranslationList('territory', $this->locale(), 2);
+	public function setSource($source) {
+		if($source) {
+			return parent::setSource($source);
+		}
 
-			// We want them ordered by display name, not country code
+		// map empty source to country list
+		// Get a list of countries from Zend
+		$source = Zend_Locale::getTranslationList('territory', $this->locale(), 2);
 
-			// PHP 5.3 has an extension that sorts UTF-8 strings correctly
-			if (class_exists('Collator') && ($collator = Collator::create($this->locale()))) {
-				$collator->asort($source);
-			}
-			// Otherwise just put up with them being weirdly ordered for now
-			else {
-				asort($source);
-			}
+		// We want them ordered by display name, not country code
 
-			// We don't want "unknown country" as an option
-			unset($source['ZZ']);
+		// PHP 5.3 has an extension that sorts UTF-8 strings correctly
+		if (class_exists('Collator') && ($collator = Collator::create($this->locale()))) {
+			$collator->asort($source);
+		} else {
+			// Otherwise just put up with them being weirdly ordered for now
+			asort($source);
 		}
```

Comment:
Many things like this should be extracted to smaller methods. Why does the constructor need to know about a global collator? How can that be tested? I mean: how would you test the field with a collator and then without?